src/ledger/database/lmdb.zig

```diff
@@ -0,0 +1,713 @@
+const std = @import("std");
+const c = @import("lmdb");
+const sig = @import("../../sig.zig");
+const database = @import("lib.zig");
+
+const Allocator = std.mem.Allocator;
+
+const BytesRef = database.interface.BytesRef;
+const ColumnFamily = database.interface.ColumnFamily;
+const IteratorDirection = database.interface.IteratorDirection;
+const Logger = sig.trace.Logger;
+
+const key_serializer = database.interface.key_serializer;
+const value_serializer = database.interface.value_serializer;
+
+pub fn LMDB(comptime column_families: []const ColumnFamily) type {
+    return struct {
+        allocator: Allocator,
+        env: *c.MDB_env,
+        dbis: []const c.MDB_dbi,
+        path: []const u8,
+
+        const Self = @This();
+
+        pub fn open(allocator: Allocator, logger: Logger, path: []const u8) anyerror!Self {
+            logger.info().log("Initializing LMDB");
+            const owned_path = try allocator.dupe(u8, path);
+
+            // create and open the database
+            const env = try ret(c.mdb_env_create, .{});
+            try result(c.mdb_env_set_maxdbs(env, column_families.len));
+            try result(c.mdb_env_open(env, @ptrCast(path), 0, 0o700));
+
+            // begin transaction to create column families aka "databases" in lmdb
+            const txn = try ret(c.mdb_txn_begin, .{ env, null, 0 });
+            errdefer c.mdb_txn_abort(txn);
+
+            // allocate cf handles
+            const dbis = try allocator.alloc(c.MDB_dbi, column_families.len);
+            errdefer allocator.free(dbis);
+
+            // save cf handles
+            inline for (column_families, 0..) |cf, i| {
+                // open cf/database, creating if necessary
+                dbis[i] = try ret(c.mdb_dbi_open, .{
+                    txn,
+                    @as([*]const u8, @ptrCast(cf.name)),
+                    c.MDB_CREATE,
+                });
```

Comment:
a nit:
- no need to ptrCast, `.ptr` on the slice will give you a `[*]const u8`

the issue:
I'm pretty sure lmdb will be expecting a c-style 0-terminated string. So giving it this pointer will not be safe - if this worked so far, it's likely only down to coincidence that the slice's data ended with a 0.
Two options:
- use a `[:0]const u8` for cf.name, which is a valid slice *and* a valid sentinel terminated pointer
- use allocator.dupeZ to copy it into a buffer that ends with a 0 

(this compiling at all is the issue with using C pointer types in general)

Reply:
> I'm pretty sure lmdb will be expecting a c-style 0-terminated string. 

I don't think so. LMDB always requires you to specify the length. It needs to use this because there is no safe sentinel value for a slice of arbitrary bytes.

Reply:
ac1e07da776c82a5ca62f1bae7590c8342d4aa1a

Reply:
I forgot the context here is actually a string, not the storage of bytes in the database. I'll fix it

Reply:
b8adf694e731c1828150f8c802193b60abcd45f7